### PR TITLE
feat(playwright): LLM-Powered Playwright Script Generation

### PR DESCRIPTION
**Pull request by ThusharaJ**
This feature adds the ability to automatically generate complete, runnable Playwright scripts from agent sessions. The implementation leverages LLM capabilities to transform agent actions into clean, maintainable playwright test automation code.

**Core functionality:**

- Extract and process agent actions (both initial and history actions)
- Transform action data into LLM-friendly format
- Generate complete Playwright scripts via LLM
- Save scripts to output directory with descriptive names
- Mask sensitive data throughout the pipeline

**Key components:**

1. Action extraction system:

- Added extract_playwright_actions method to process agent actions
- Support for initial navigation actions with include_initial_actions parameter
- Proper formatting of action parameters and results
- Conditional JSON output based on debug logging level

2. LLM integration:

- Designed prompt template for Playwright code generation
- Implemented context handling for complex action sequences
- Added support for headless/headed mode configuration
- Ensured sensitive data masking with tags

3. Script management:

- Automatic script naming based on actions and timestamps
- Standardized output directory structure
- Proper file permissions and error handling
- Simplified path handling for better user experience

4. Example implementation:

- Created comprehensive example in playwright-script-generator.py
- Added asyncio handling with proper timeouts and cleanup
- Demonstrated end-to-end workflow from agent creation to script execution
- Included initial actions for complete page navigation

5. Comprehensive test suite:

- Added 12 test cases covering all aspects of the feature
- Test cases for basic action extraction
- Dedicated tests for initial actions handling
- Tests for empty history edge cases
- Tests for debug mode JSON output
- Tests for script generation with various inputs
- Tests for LLM prompt construction
- Tests for script cleanup and path generation
- All tests are async-compatible and use proper mocking

**Technical implementation details:**

- Used Anthropic LLM for script generation via langchain integration
- Implemented proper async/await patterns throughout
- Added signal handling for graceful termination
- Used timeout protection to prevent hanging
- Ensured proper resource cleanup even during interruption
- Maintained backward compatibility with existing agent functionality
- Added type hints and docstrings for better code quality

This feature enables users to easily convert their browser-use agent sessions into reusable Playwright automation scripts, significantly enhancing the tool's utility for testing and automation workflows.

    

---

## Summary by cubic
Added automatic Playwright script generation from agent sessions using LLMs, allowing users to convert browser actions into runnable test scripts.

- **New Features**
  - Extracts agent actions and transforms them into Playwright scripts with LLM integration.
  - Saves scripts with descriptive names and masks sensitive data.
  - Includes an example workflow and a full async-compatible test suite.

- **Dependencies**
  - Added aiofiles for async file operations.

### CODE REVIEW

**Comment by ThusharaJ** [comment]
@MagMueller Hi! Just wanted to check if this PR needs any changes. Happy to revise if needed. Appreciate a review when time permits. 🙏

File: browser_use/agent/service.py [hunk 1]
```diff
@@ -1868,6 +1868,207 @@ async def close(self):
...
+
+		# Clean up the script to remove any markdown formatting that the LLM might have included
+		# LLMs often return code in markdown code blocks with ```python and ``` delimiters
+		logger.debug('Cleaning up generated script content')
+		if script_content.startswith('```python'):
+			script_content = script_content.split('```python', 1)[1]
+		elif script_content.startswith('```'):
+			script_content = script_content.split('```', 1)[1]
+		if script_content.endswith('```'):
+			script_content = script_content.rsplit('```', 1)[0]
```
Comment (servusdei2018) [inline browser_use/agent/service.py:2050]:
```diff
- if script_content.startswith('```python'):
-     script_content = script_content.split('```python', 1)[1]
- elif script_content.startswith('```'):
-     script_content = script_content.split('```', 1)[1]
- if script_content.endswith('```'):
-     script_content = script_content.rsplit('```', 1)[0]
+ script_content = script_content.replace('```python', '').replace('```', '')
```

This could be a oneliner, and prevents a possible `IndexError`.

File: browser_use/agent/service.py [hunk 2]
```diff
@@ -1868,6 +1868,207 @@ async def close(self):
...
+		if output_path:
+			path_obj = Path(output_path)
+			# Ensure the output directory exists
+			path_obj.parent.mkdir(parents=True, exist_ok=True)
+
+			# Write the script to the file with UTF-8 encoding using aiofiles for async I/O
+			import aiofiles
+
+			async with aiofiles.open(path_obj, 'w', encoding='utf-8') as f:
+				await f.write(script_content)
```
Comment (servusdei2018) [inline browser_use/agent/service.py:2066]: I don't think this use case is enough justification to introduce a new `aiofiles` dependency.

File: browser_use/agent/service.py [hunk 3]
```diff
@@ -1868,6 +1868,207 @@ async def close(self):
...
+	def extract_playwright_actions(self, output_path: str | Path = None, include_initial_actions: bool = True) -> list[dict]:
+		"""
+		Extract Playwright actions from the agent's history and save as JSON for developer testing.
+
+		Args:
+			output_path: Optional path to save the extracted actions as JSON
+			include_initial_actions: Whether to include initial actions in the extracted list
+
+		Returns:
+			List of action dictionaries with action_name, params, and result
```
Comment (servusdei2018) [inline browser_use/agent/service.py:1880]: Should probably be its own class.

**Comment by pirate** [comment]
we just removed this feature actually, we're focusing all our effort around runnable scripts in: https://github.com/browser-use/workflow-use

You should contribute this over there instead!

File: browser_use/agent/service.py (hunk 1 above)
Comment (ThusharaJ) [reply browser_use/agent/service.py:2050]: @servusdei2018 thanx for the comment.

File: browser_use/agent/service.py (hunk 2 above)
Comment (ThusharaJ) [reply browser_use/agent/service.py:2066]: @servusdei2018 The initial version of the code was synchronous, but the CI tests were failing, so it had to be updated to use asyncio.

File: browser_use/agent/service.py (hunk 3 above)
Comment (ThusharaJ) [reply browser_use/agent/service.py:1880]: @servusdei2018 Thanx for the comment.

**Comment by ThusharaJ** [comment]
@pirate Thanks for letting me know. I'll take a look at it.
